### PR TITLE
Make `contents` API scale

### PR DESCRIPTION
**Pull request by dbutenhof**
PBENCH-1321

The `/datasets/{id}/contents` API includes into several unexpectedly expensive steps:

1. Finding the tarball (by MD5 value) within the `ARCHIVE` tree using a `glob`
2. Fully discovering all tarballs within the controller directory
3. Unpacking the tarball into a cache directory using `tar`
4. Building a "map" of the contents of the unpacked tarball subtree

This PR includes mitigations for all but the `tar` unpack step:

1. Use the `server.tarball-path` metadata instead of searching the disk
2. Only discover the target tarball rather than the entire controller
3. Skip the "map" and evaluate the actual target path within the cache

Finding a tarball within our 30Tb `ARCHIVE` tree can take many minutes, while identifying the controller directory from the tarball path takes a fraction of a second.

Depending on the number of tarballs within a controller (some have many), full controller discovery has been observed to take half a minute; while populating only the target tarball takes a fraction of a second.

Building the map for a large tarball tree can take minutes, whereas discovery of the actual relative file path within the cache runs at native (Python) file system speeds.

### CODE REVIEW

**Comment by dbutenhof** [comment]
Bad coverage means more test cases. And more test cases means ... more edge cases to fix. 😆

File: lib/pbench/server/cache_manager.py [hunk 1]
```diff
@@ -1393,16 +1530,35 @@ def find_dataset(self, dataset_id: str) -> Tarball:
...
+        sql = start
+        try:
+            dataset = Dataset.query(resource_id=dataset_id)
+            sql = time.time()
+            tarball = Path(Metadata.getvalue(dataset, Metadata.TARBALL_PATH))
+            controller_dir = tarball.parent
+            if controller_dir.name == dataset_id:
+                controller_dir = controller_dir.parent
+        except Exception as e:
+            raise TarballNotFound(dataset_id) from e
```
Comment (webbnh) [inline lib/pbench/server/cache_manager.py:1545]: Which of these statements is likely to raise an exception?  (The first one, obviously...but, are there any others?  Should some or all of the remainder be moved out of the `try` block?)

File: lib/pbench/server/cache_manager.py [hunk 2]
```diff
@@ -1393,16 +1530,35 @@ def find_dataset(self, dataset_id: str) -> Tarball:
...
+        found = time.time()
+        controller = self.controllers.get(
+            controller_dir.name, self._add_controller(controller_dir, False)
+        )
```
Comment (webbnh) [inline lib/pbench/server/cache_manager.py:1549]: Oooh, this is clever!  (I'll have to remember this trick.)

File: lib/pbench/test/unit/server/test_cache_manager.py [hunk 3]
```diff
@@ -1552,6 +1553,152 @@ def read(self, size: int = -1) -> bytes:
...
+
+        base: Path = cm.find_dataset(md5).unpacked
+        subdir = base / "subdir"
+        subdir.mkdir()
+        ldir = base / "dir_link"
+        ldir.symlink_to(subdir)
+        lfile = base / "file_link"
+        lfile.symlink_to(Path(base / "metadata.log"))
+        badlink = base / "bad_link"
+        badlink.symlink_to("/etc/passwd")
```
Comment (webbnh) [inline lib/pbench/test/unit/server/test_cache_manager.py:1601]: This is a symlink with an absolute path; would it make sense to have a bad symlink with a relative path (i.e., an attempt to "walk up" out of the tarball)?

File: lib/pbench/test/unit/server/test_cache_manager.py [hunk 4]
```diff
@@ -1552,6 +1553,152 @@ def read(self, size: int = -1) -> bytes:
...
+                    "link_type": "OTHER",
+                    "uri": f"{root}/inventory/fifo_link",
+                },
+                {
+                    "name": "file_link",
+                    "type": "SYMLINK",
+                    "link": "metadata.log",
+                    "link_type": "FILE",
+                    "uri": f"{root}/inventory/metadata.log",
+                },
```
Comment (webbnh) [inline lib/pbench/test/unit/server/test_cache_manager.py:1649]: Any interest in a symlink to a symlink?  (Probably best not, because then we would have to decide whether the target symlink is linked to a directory, a file, broken, etc. .... 😉)

File: lib/pbench/test/unit/server/test_cache_manager.py [hunk 5]
```diff
@@ -1580,6 +1733,8 @@ def test_find(
         assert tarball
         assert tarball.name == dataset_name
         assert tarball.resource_id == md5
+        monkeypatch.setattr(Dataset, "query", mock_query)
+        monkeypatch.setattr(Metadata, "getvalue", lambda d, k: tarball.tarball_path)
```
Comment (webbnh) [inline lib/pbench/test/unit/server/test_cache_manager.py:1737]: Looking at the corresponding mock in `lib/pbench/test/unit/server/test_datasets_compare.py`, this `getvalue()` mock looks over-simplified/generalized:  for one thing, it's missing the third, optional parameter (`user`); and, for another, it doesn't do any kind of check on the requested key value (which seems like it could produce "interesting" and mysterious results...).

**Review by webbnh — DISMISSED** [review]
Looks good; however, I have the usual collection of nits, pointed questions, and suggestions.

File: lib/pbench/test/unit/server/test_cache_manager.py (hunk 3 above)
Comment (dbutenhof) [reply lib/pbench/test/unit/server/test_cache_manager.py:1601]: I did at one point. I converted to an absolute path because `Path.absolute()` doesn't drop a relative `..` and `relative_to` just drops everything up to the `..`, meaning we're silently going outside the boundaries. I figured out a solution by going to `os.path.abspath`, which *does* drop the `..` so that `relative_to` will fail ... but apparently forgot to reinstate the test case. 😦

File: lib/pbench/test/unit/server/test_cache_manager.py (hunk 4 above)
Comment (dbutenhof) [reply lib/pbench/test/unit/server/test_cache_manager.py:1649]: Actually, since we use `resolve`, intermediate links should be completely transparent to us. I'm unconvinced that it'd be worth the effort of writing our own `readlink` loop. In theory, there's a hole here if we were to get a symlink *outside* the cache directory to a symlink back *into* the cache directory. One could argue it's technically illegal, but we'd never know. Since it's a controlled server environment it's also unlikely enough I find it hard to care much.

Still ... that's an interesting test case. (Not the mad dash out and back in, but an indirect link ...)

File: lib/pbench/test/unit/server/test_cache_manager.py (hunk 5 above)
Comment (dbutenhof) [reply lib/pbench/test/unit/server/test_cache_manager.py:1737]: It's not more complicated than needed here.

**Review by dbutenhof — COMMENTED** [review]
I made some of the changes, including some new test links.

I also got an idea this afternoon (which is why I'm working way late today!) about improving the report generator performance by extending my `find_dataset` changes to an alternate "full discovery" by pulling all the dataset resource IDs and `tarball-path` values from SQL instead of `iterdir`-ing through the `ARCHIVE`s. Turns out that cuts cache discovery from nearly 2 hours to about 25 minutes.

File: lib/pbench/server/cache_manager.py [hunk 6]
```diff
@@ -1423,16 +1431,22 @@ def __init__(self, options: PbenchServerConfig, logger: Logger):
...
+
+        Returns:
+            CacheManager instance for chaining
         """
-        self._discover_controllers()
+        if full:
+            self._discover_controllers()
+        else:
+            self._discover_datasets()
         return self
```
Comment (webbnh) [inline lib/pbench/server/cache_manager.py:1450]: Why not just replace
```
    cache_m.full_discovery(full=...)
```
with
```
    cache_m._discover_datasets()
```
or
```
    cache_m._discover_controllers()
```
as appropriate (with non-underscore-prefixed names, perhaps) rather than wrapping them?

File: lib/pbench/server/cache_manager.py [hunk 7]
```diff
@@ -1538,11 +1604,11 @@ def find_dataset(self, dataset_id: str) -> Tarball:
...
             sql = time.time()
             tarball = Path(Metadata.getvalue(dataset, Metadata.TARBALL_PATH))
-            controller_dir = tarball.parent
-            if controller_dir.name == dataset_id:
-                controller_dir = controller_dir.parent
         except Exception as e:
             raise TarballNotFound(dataset_id) from e
+        controller_dir = tarball.parent
+        if controller_dir.name == dataset_id:
+            controller_dir = controller_dir.parent
```
Comment (webbnh) [inline lib/pbench/server/cache_manager.py:1611]: Alternately,
```python
        tbp = tarball.parent
        controller_dir = tbp.parent if tbp.name == dataset_id else tbp
```
Ditto lines 1561-1563.

File: lib/pbench/server/cache_manager.py [hunk 8]
```diff
@@ -1507,6 +1521,58 @@ def _discover_controllers(self):
...
+                Dataset.resource_id,
+                Metadata.value["tarball-path"].as_string(),
+            )
+            .execution_options(stream_results=True)
+            .outerjoin(
+                Metadata,
+                and_(Dataset.id == Metadata.dataset_ref, Metadata.key == "server"),
+            )
+            .yield_per(500)
+            .all()
```
Comment (webbnh) [inline lib/pbench/server/cache_manager.py:1544]: Did you look at using a number larger than `500`?  Given the size of a "row" and the number of rows that there might be, it seems like 10 or 20 times that many would be better.

File: lib/pbench/server/cache_manager.py [hunk 9]
```diff
@@ -1507,6 +1521,58 @@ def _discover_controllers(self):
...
+        )
+
+        for name, resource_id, path in rows:
+            if not all((resource_id, path)):
```
Comment (webbnh) [inline lib/pbench/server/cache_manager.py:1548]: Why check for the `resource_id` but not the `name`?  (It seems like the only item at question is the `path`, so why not check just that?)

File: lib/pbench/server/cache_manager.py [hunk 10]
```diff
@@ -799,6 +800,131 @@ def extract(tarball_path: Path, path: str) -> Inventory:
...
+                            if target.is_dir():
+                                uri = f"{origin}/contents/{link}"
+                                type = CacheType.DIRECTORY
+                                append_to = dir_list
+                            elif target.is_file():
+                                uri = f"{origin}/inventory/{link}"
+                                type = CacheType.FILE
+                            else:
+                                uri = f"{origin}/inventory/{relative}"
+                                type = CacheType.OTHER
```
[the remark this answers is not in the document]
Comment (webbnh) [reply lib/pbench/server/cache_manager.py]: I concur on the `BROKEN` case...but why is that appropriate for the `OTHER` case?  (In the `OTHER` case, we successfully resolved the symlink to a directory entry accessible within the tarball via a relative path, so won't `f"{origin}/inventory/{link}"` and `f"{origin}/inventory/{relative}" address the same file, with the latter just taking a more circuitous route, if the route is different at all?)

File: lib/pbench/server/cache_manager.py [hunk 11]
```diff
@@ -799,6 +800,131 @@ def extract(tarball_path: Path, path: str) -> Inventory:
...
+            except Exception:
+                raise CacheExtractError(self.name, path)
+            if artifact.is_dir() and not artifact.is_symlink():
+                dir_list = []
+                file_list = []
+                for f in artifact.iterdir():
+                    relative = f.relative_to(self.unpacked)
+                    append_to = file_list
+                    if f.is_symlink():
+                        target = f.resolve()
```
[the remark this answers is not in the document]
Comment (webbnh) [reply lib/pbench/server/cache_manager.py:832]: It is?  I see the `append()` call, but it's inside the `if` block...I _think_....  (The alternate value for `append_to` is assigned in the `else` clause of the `try` block inside the `if` block.)

File: lib/pbench/server/cache_manager.py [hunk 12]
```diff
@@ -799,6 +800,131 @@ def extract(tarball_path: Path, path: str) -> Inventory:
...
+                        try:
+                            link = target.relative_to(self.unpacked)
+                        except Exception:
+                            link = f.readlink()
+                            uri = f"{origin}/inventory/{relative}"
+                            type = CacheType.BROKEN
```
[the remark this answers is not in the document]
Comment (webbnh) [reply lib/pbench/server/cache_manager.py]: Yes, but Python lets us do lots of things, at least some of which we really shouldn't!  🙃

File: lib/pbench/test/unit/server/test_cache_manager.py (hunk 4 above)
Comment (webbnh) [reply lib/pbench/test/unit/server/test_cache_manager.py:1654]: I don't think I care about the jump-out-jump-in case:  for starters, we end up back inside the unpacked tarball, so presumably there is no "leak"; beyond that, it seems unlikely in the extreme that the jump-in symlink would ever exist (and, if it does, we have bigger problems!).

File: lib/pbench/test/unit/server/test_cache_manager.py (hunk 5 above)
Comment (webbnh) [reply lib/pbench/test/unit/server/test_cache_manager.py:1756]: The fact that it has the "wrong" number of parameters but still works is just fascinating.  I so love Python....

**Review by webbnh — APPROVED** [review]
The updates look generally good.  I've got follow-ups on two of my comments from the first pass (the rest are resolved -- thanks!), and I've got a few thoughts on your revisions.  The foremost is that making `CacheManager.full_discovery()` into a wrapper seems like it causes more problems than it really solves.  (The result is good, but the approach has some flaws.)  The others are smaller bits, nits, and suggestions.

File: lib/pbench/cli/server/report.py [hunk 13]
```diff
@@ -475,7 +475,7 @@ def report(
             cache_m = CacheManager(config, logger)
             verifier.status("starting discovery")
             watcher.update("discovering cache")
-            cache_m.full_discovery()
+            cache_m.full_discovery(full=False)
```
[the remark this answers is not in the document]
Comment (dbutenhof) [reply lib/pbench/cli/server/report.py:478]: Both varieties are "full" except in the presence of prior server errors. The new mode is faster and will ignore "orphaned" tarballs in the ARCHIVE tree that don't represent SQL `Dataset`s. The modern variation of the old "audit" tool would definitely want to locate and diagnose these, but in the context of my report generator the faster "full discover" is much more useful.

I suppose I could call it "paranoid" or "exhaustive" or just "slow", but I frankly didn't spend much time worrying about it and I'm not really inclined to start now.

File: lib/pbench/server/cache_manager.py (hunk 8 above)
Comment (dbutenhof) [reply lib/pbench/server/cache_manager.py:1544]: I used these parameters to stream a query in the report generator earlier, and I just copied them. The intention is to avoid overloading Python's heap with hundreds of thousands of rows at once, but I really did't want to get into trying to figure out how to get good enough low-level usage data to seriously optimize it. Yeah, it's possible there'd even be a visible difference at the CLI level, which is the only place this is used: but I'm not sure that investigation is a particularly good use of my time right now.

File: lib/pbench/server/cache_manager.py (hunk 9 above)
Comment (dbutenhof) [reply lib/pbench/server/cache_manager.py:1548]: The name is only used for a log message, but both the `resource_id` and `path` are critical. Yeah, technically we can't have a `Dataset` without a `resource_id` (the column isn't nullable), but as long as I was checking I figured I'd check both just because the code looked odd checking just one.

File: lib/pbench/server/cache_manager.py (hunk 6 above)
Comment (dbutenhof) [reply lib/pbench/server/cache_manager.py:1450]: From the outside, this is two subtly different styles of full discovery, and I wanted to package them that way instead of inventing *two* new names for external entry points.

File: lib/pbench/server/cache_manager.py (hunk 7 above)
Comment (dbutenhof) [reply lib/pbench/server/cache_manager.py:1611]: Ugh. Ugly ternary alert ... ☹️

File: lib/pbench/cli/server/report.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (webbnh) [reply lib/pbench/cli/server/report.py:478]: > Both varieties are "full"

OK, so that justifies keeping the function name, but it makes the keyword name look silly.

I would be inclined to replace `full` with `directory_search` or its opposite, `use_database`; or, perhaps, `by_controller` vs. `by_dataset`.

File: lib/pbench/server/cache_manager.py (hunk 7 above)
Comment (webbnh) [reply lib/pbench/server/cache_manager.py:1611]: Ugly is a bit judgey!  You want to assign one of two values to `controller_dir`, that just cries out for a ternary!!  (An `if` statement which overwrites `controller_dir`, now that's a candidate for "ugly"....  🙂)

File: lib/pbench/server/cache_manager.py (hunk 8 above)
Comment (webbnh) [reply lib/pbench/server/cache_manager.py:1544]: > I'm not sure that investigation is a particularly good use of my time right now.

Prob'ly not...I was just thinking that 130K results taken 500 at a time is a lot of takes, especially since I suspect that each block of 500 doesn't take up much space in real terms.

File: lib/pbench/server/cache_manager.py (hunk 9 above)
Comment (webbnh) [reply lib/pbench/server/cache_manager.py:1548]: Unfortunately, the code also looks odd checking two (I had to think carefully about how the `not` applied to `any()` and why there was an "extra" set of parens in the argument list...all of which go away if you only check `path`).

File: lib/pbench/cli/server/report.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (dbutenhof) [reply lib/pbench/cli/server/report.py]: I actually renamed the parameter to `search` (vs "by query", I guess).

**Review by webbnh — APPROVED** [review]
Looks good.

Are you contemplating acting on or responding to the two remaining open conversations ([1](https://github.com/distributed-system-analysis/pbench/pull/3609#discussion_r1496230268), [2](https://github.com/distributed-system-analysis/pbench/pull/3609#discussion_r1496180826)) or should I just close them?

File: lib/pbench/server/cache_manager.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (dbutenhof) [reply lib/pbench/server/cache_manager.py]: That's actually a comment for `inventory` rather than `contents`. We don't want to expose the real target of an `OTHER` file (e.g., a fifo or whatever), which is why we use `OTHER`. Yeah, it's possible that `inventory` will transparently resolve the link and return the `fifo`, which I suspect we really don't want. I'm not going to mess with that here.

File: lib/pbench/server/cache_manager.py [hunk 14]
```diff
@@ -911,17 +913,17 @@ def get_contents(self, path: str, origin: str) -> JSONOBJECT:
...
                 val = {
                     "name": artifact.name,
-                    "type": type.name,
+                    "type": link_type.name,
                     "uri": f"{origin}/{access}/{arel}",
                 }
-                if type is CacheType.SYMLINK:
+                if link_type is CacheType.SYMLINK:
                     val["link"] = str(trel)
                     val["link_type"] = ltype.name
```
[the remark this answers is not in the document]
Comment (dbutenhof) [reply lib/pbench/server/cache_manager.py]: I guess I missed this one. There were too many name contexts, and some probably aren't ideal; but I don't think it's worth messing with now.

**Review by dbutenhof — COMMENTED** [review]
Apparently I have several responses written earlier this afternoon that didn't get posted. Huh. You can enjoy them at your leisure.
